C/jets.c

```diff
@@ -1043,6 +1043,165 @@ DIVIDES_(16)
+  *q = estQ <= UINT32_MAX ? estQ : UINT32_MAX;
+
+  /* *q * bh <= estQ * bh <= ah */
+  uint_fast64_t rh = ah - 1u * *q * bh;
+  uint_fast64_t d = 1u * *q * bl;
+
+  /* Test if our estimated remainder, A - *q * B is negative.
+   * A - *q * B  == [ah;al] - *q * [bh;bl]
+   *             == [ah - *q * bh;al - *q * bl]
+   *             == [rh;al] - d
+   *
+   * This value is negative when [rh;al] < d.
+   * Note that d is 64 bit and thus if rh is greater than UINT32_MAX, then this value cannot be negative.
+   */
+  /* This loop is exectued at most twice. */
+  while (rh <= UINT32_MAX && 0x100000000u*rh + al < d) {
+    /* Our estimated remainder, A - *q * B is negative. */
+    /* 0 < d == *q * bl and hence 0 < *q, so this decrement does not underflow. */
+    (*q)--;
+    rh += bh;
+    /* rh == ah - *q * bh */
+    d -= bl;
+    /* d == *q * bl */;
+  }
+  /* *q is the correct quotient. */
+
+  /* Compute the remainder.
+   * The computation below is performed modulo 2^64.
+   * However since we know the remainder lies within [0,b) which is within [0,2^64),
+   * the computation must end up with the correct result even if overflow/underflow occurs.
+   */
+  *r = 0x100000000u*rh + al - d;
+}
+
+/* div_mod_128_64 : TWO^128 * TWO^64 |- TWO^64 * TWO^64 */
+bool div_mod_128_64(frameItem* dst, frameItem src, const txEnv* env) {
+  (void) env; /* env is unused. */
+  uint_fast32_t qh, ql;
+  uint_fast64_t r;
+  uint_fast64_t ah = read64(&src);
+  uint_fast32_t am = read32(&src);
+  uint_fast32_t al = read32(&src);
+  uint_fast64_t b = read64(&src);
+
+  /* div2n1n is only defined when 2^(n-1) <= b and when the quotient q < 2^n. */
+  if (0x8000000000000000 <= b && ah < b) {
+    /* We compute the division and remainder of A = [[ah;am];al] by B = b by long division with 32-bit "digits"
+     *
+     * 1.       Q
+     *      ______
+     *    BB) AAAA
+     *        XXX
+     *        ---
+     *         RR
+     *
+     * First divide the high 3 "digit"s (96-bits) of A by the two "digit"s (64-bits) of B,
+     * returning the first "digit" (high 32-bits) of the quotient, and an intermediate remainer consisiting of 2 "digit"s (64-bits).
+     */
+    div_mod_96_64(&qh, &r, ah, am, b);
+    simplicity_debug_assert(r < b);
+    /* 2.       QQ
+     *      ______
+     *    BB) AAA|
+     *        XXX|
+     *        ---|
+     *         RRA
+     *         XXX
+     *         ---
+     *          RR
+     *
+     * Then append the last "digit" of A to the intermidiate remainder and divide that value (96_bits) by the two "digit"s (64-bits) of B,
+     * returning the second "digit" (low 32-bits) of the quotient, and the final remainer consisiting of 2 "digit"s (64-bits).
+     */
+    div_mod_96_64(&ql, &r, r, al, b);
+    write32(dst, qh);
+    write32(dst, ql);
+    write64(dst, r);
+  } else {
+    /* Set all the bits in the output when the input is out of bounds. */
```

Comment:
In d80cc978a89103d5a5717a1b105a010d28d02e06:

Why did you choose to this rather than returning an optional type?

Reply:
I guess, the idea is that this is a primitive and if you call this with wrong values that's a programming error, and we don't want to force the caller to destructure things in that case?

You could just `fail` in this case.

Reply:
Yeah.  The main disadvantage with using `fail` is that the term is no longer in the Core (assertion free) subset.  It's hard to tell what is best here.

Reply:
I don't go with an optional type because this is an expression to recursively build up division operators for wider and wider integers, and adding code to handle possible failure cases for inputs that one knows cannot fail is somewhat annoying.

Also worth point out that the remainder value returned is always larger than than what can be returned during a successful run.